webknossos-datastore/app/com/scalableminds/webknossos/datastore/explore/NgffExplorationUtils.scala

```diff
@@ -0,0 +1,217 @@
+
+  protected def magFromTransforms(coordinateTransforms: List[NgffCoordinateTransformation],
+                                  voxelSizeInAxisUnits: Vec3Double,
+                                  axisOrder: AxisOrder)(implicit ec: ExecutionContext): Fox[Vec3Int] = {
+    def isPowerOfTwo(x: Int): Boolean =
+      x != 0 && (x & (x - 1)) == 0
+
+    val combinedScale = extractAndCombineScaleTransforms(coordinateTransforms, axisOrder)
+    val mag = (combinedScale / voxelSizeInAxisUnits).round.toVec3Int
+    for {
+      _ <- bool2Fox(isPowerOfTwo(mag.x) && isPowerOfTwo(mag.y) && isPowerOfTwo(mag.z)) ?~> s"invalid mag: $mag. Must all be powers of two"
+    } yield mag
+  }
+
+  protected def ensureElementClassForSegmentationLayer(e: ElementClass.Value): ElementClass.Value =
+    e match {
+      case ElementClass.int8  => ElementClass.uint8
+      case ElementClass.int16 => ElementClass.uint16
+      case ElementClass.int32 => ElementClass.uint32
+      case ElementClass.int64 => ElementClass.uint64
+      case _                  => e
+    }
+
+  /**
+    * Guesses the voxel size from all transforms of an ngff multiscale object.
+    * Note: the returned voxel size is in axis units and should later be combined with those units
+    *   to get a webknossos-typical voxel size in nanometers.
+    * Note: allCoordinateTransforms is nested: the inner list has all transforms of one ngff “dataset” (mag in our terminology),
+    *   the outer list gathers these for all such “datasets” (mags) of one “multiscale object” (layer)
+    */
+  protected def extractVoxelSizeInAxisUnits(allCoordinateTransforms: List[List[NgffCoordinateTransformation]],
+                                            axisOrder: AxisOrder)(implicit ec: ExecutionContext): Fox[Vec3Double] = {
+    val scales = allCoordinateTransforms.map(t => extractAndCombineScaleTransforms(t, axisOrder))
+    val smallestScaleIsUniform = scales.minBy(_.x) == scales.minBy(_.y) && scales.minBy(_.y) == scales.minBy(_.z)
+    for {
+      _ <- bool2Fox(smallestScaleIsUniform) ?~> "ngff scales do not agree on smallest dimension"
+      voxelSizeInAxisUnits = scales.minBy(_.x)
+    } yield voxelSizeInAxisUnits
+  }
+
+  private def extractAndCombineScaleTransforms(coordinateTransforms: List[NgffCoordinateTransformation],
+                                               axisOrder: AxisOrder): Vec3Double = {
+    val filtered = coordinateTransforms.filter(_.`type` == "scale")
+    val scalesFromTransforms = filtered.flatMap(_.scale)
+    val xFactors = scalesFromTransforms.map(_(axisOrder.x))
+    val yFactors = scalesFromTransforms.map(_(axisOrder.y))
+    val zFactors = if (axisOrder.hasZAxis) scalesFromTransforms.map(_(axisOrder.zWithFallback)) else Seq(1.0, 1.0)
+    Vec3Double(xFactors.product, yFactors.product, zFactors.product)
+  }
+
+  protected def getShape(dataset: NgffDataset, path: VaultPath): Fox[Array[Int]] = ???
```

Comment:
can the `= ???` be removed? Then the compiler should check that all classes implementing the trait implement this.